snuba/datasets/schemas/__init__.py

```diff
@@ -107,10 +117,32 @@ def get_column_differences(self, expected_columns: Mapping[str, str]) -> List[st
                 continue
 
             expected_type = self.get_columns()[column_name].type.for_schema()
+
+            column_type = self.__get_column_type(column_desc)
             if column_type != expected_type:
                 errors.append(
                     "Column '%s' type differs between local ClickHouse and schema! (expected: %s, is: %s)"
                     % (column_name, expected_type, column_type)
                 )
 
         return errors
+
+    def __get_column_type(self, column: MigrationSchemaColumn):
+        (column_type, default_type, default_expr, codec_expr) = column
+
+        column_type_expr = column_type
+
+        if default_type and default_expr:
+            # Strip the cast function if there is one
+            DEFAULT_EXPR_RE = re.compile(r"^CAST\((.*), (.*)\)$", re.IGNORECASE)
+            match = DEFAULT_EXPR_RE.match(default_expr)
+            if match:
+                default_expr = match.groups()[0]
+            column_type_expr = column_type_expr + " {} {}".format(
+                default_type, default_expr
+            )
+
+        if codec_expr:
+            column_type_expr = "{} CODEC ({})".format(column_type_expr, codec_expr)
+
```

Comment:
Our migration system is so broken that it is hard to give a constructive feedback on how this should look like without 3 months of refactoring, still I think we can make a step in the right direction by start using ColumnType classes and subclasses instead of comparing strings.

The ColumnType abstraction is far from being a good one to manage migrations, but it is already present in the schema of the column and it would give you some structure at least.
So I suggest that, instead of fetching the ColumnType for the column from the schema, formatting it into string and comapring with a tweaked string coming from the `DESCRIBE` call, you parse the string coming from DESCRIBE into a ColumnType (that can be nested) and then compare it with the one present in the schema. 

Since this is not time critical now, I think it would be a good investment.

Reply:
Tackled this first since I think having a reliable way to compare current database state with the schema of a dataset will be generally useful.

I think using our standard ColumnTypes (and doing away with MigrationSchemaColumn completely) will be good. Probably don't need a separate column type representation just for migration.

It seems like it will be quite a bit harder than comparing the strings though, since:
- We need to store another set of mappings to go from string column descriptions back into ColumnType structures. Going the other way is easier because ColumnType already implements the `for_schema()` method for this.
- There might be more than one way to represent columns in our schema, so it's ambiguous what the schema should actually be
- One case I can think of that is like this is our Nested column type where we use the `.` notation as a convention but it the same output can possibly be represented in multiple ways in the dataset

Reply:
It seems both your first and second point apply equally to string comparison int hat you have to tweak the representations exactly in this method. I think we should think this problem through before adding a hack to avoid the warnings. 
Is there anything that would be blocked on this PR?

Reply:
No blockers. What do you mean that they apply equally though? My point was the opposite; I don't think those points apply to string comparison at all.
- Mapping from column type to string already exists (this is `for_schema()`), mapping from string to ColumnType needs to be added and is only needed if we are comparing the actual ColumnType instances
- There is a 1:many relationship between ColumnType: string output. Columns defined in schema always yield the same string in Clickhouse. The opposite isn't true, given a `DESCRIBE TABLE` output there are many ways it can be represented in the schema. (Nested vs Array is the one example of this)

Reply:
One other thing that is not very clear is how column type order is supposed to affect the query. For example is `WithCodec(WithDefault(x))` supposed to yield the same result as `WithDefault(WithCodec(x))`? Is one of these supposed to be invalid? Or are they meant to be smart enough to figure out what the right statement should be regardless of order applied?

Reply:
Last commit in this branch is the version with ColumnType and removed the MigrationSchemaColumn, there's a bit more code involved here

Reply:
```
For example is WithCodec(WithDefault(x)) supposed to yield the same statement as WithDefault(WithCodec(x))
```

That's a good question. I am not sure, could you please verify if clickhouse expects one form only ?

Reply:
I tried both and only the first variant - `WithCodecs(WithDefault(x, def), codecs)`ends up getting formatted correctly by Clickhouse. But I guess my point wasn't about what happens today (which I'd expect to be not 100% correct) but what we should be doing if a user provides the reverse `WithDefault(withCodecs(x))`. Should we throw an error immediately? Or just accept it and transform it into the right thing? What if there are other examples out there that are both accepted by Clickhouse? Should we always enforce one "correct" order to apply the column types?

Reply:
```
but what we should be doing if a user provides the reverse WithDefault(withCodecs(x)). Should we throw an error immediately? 
```
That's my question: does clickhouse accept both ? And if it accepts both does it consider them equivalent in a DESCRIBE TABLE statement ?

Reply:
Clickhouse only accepts one. I suppose they are equivalent in that they are completely separate columns in the describe table output